saleor/dashboard/category/forms.py

```diff
@@ -0,0 +1,16 @@
+from django import forms
+from django.utils.translation import ugettext_lazy as _
+
+from ...product.models import Category
+
+
+class CategoryForm(forms.ModelForm):
+    class Meta:
+        model = Category
+        exclude = []
+
+    def clean_parent(self):
+        parent = self.cleaned_data['parent']
+        if parent == self.instance:
```

Comment:
That won't suffice. You need to check whether the current object is not within its new parents ancestors. Loops in tree structures are not allowed.